tempesta_fw/http.c

```diff
@@ -615,35 +634,212 @@ tfw_http_adjust_resp(TfwHttpResp *resp, TfwHttpReq *req)
+			continue;
+		}
+		if (!tfw_connection_nfo(srv_conn))
+			break;
+		if (tfw_connection_send(srv_conn, (TfwMsg *)req, false)) {
+			list_move_tail(&req->msg.fwd_list, &err_queue);
+			continue;
+		}
+		TFW_INC_STAT_BH(clnt.msgs_forwarded);
+		if (req->flags & TFW_HTTP_SYNC_REQ)
+			break;
+	}
+	if (req == end)
+		srv_conn->flags &= ~TFW_CONN_FWD_HOLD;
+	spin_unlock_bh(&srv_conn->msg_qlock);
+
+	/* Requests from dead client connections. */
+	list_for_each_entry_safe(req, tmp, &zap_queue, msg.fwd_list) {
+		list_del_init(&req->msg.fwd_list);
+		if (!list_empty_careful(&req->msg.seq_list)) {
+			spin_lock_bh(&req->conn->msg_qlock);
+			list_del_init(&req->msg.seq_list);
+			spin_unlock_bh(&req->conn->msg_qlock);
+		}
+		tfw_http_conn_cli_dropfree((TfwHttpMsg *)req);
+	}
+
+	/* Requests that couldn't be forwarded due to an error. */
+	list_for_each_entry_safe(req, tmp, &err_queue, msg.fwd_list) {
+		list_del_init(&req->msg.fwd_list);
+		tfw_http_send_500((TfwHttpMsg *)req);
+		TFW_INC_STAT_BH(clnt.msgs_otherr);
+	}
+}
+
+/*
+ * Forward responses to the client in the correct order.
+ * Can be called in both SoftIRQ and kernel thread contexts.
+ */
+static void
+tfw_http_resp_forward(TfwHttpReq *req, TfwHttpResp *resp, bool unref_data)
+{
+	TfwHttpReq *tmp;
+	TfwConnection *cli_conn = req->conn;
+	struct list_head out_queue, *seq_queue = &cli_conn->msg_queue;
+
+	TFW_DBG3("%s: req = %p, resp = %p, unref = %d\n",
+		 __func__, req, resp, unref_data);
+
+	INIT_LIST_HEAD(&out_queue);
+	if (unref_data)
+		resp->flags |= TFW_HTTP_UNREF_DATA;
+	req->resp = (TfwHttpMsg *)resp;
+
+	spin_lock_bh(&cli_conn->msg_qlock);
+	/*
+	 * It the list is empty, then it's either a bug, or the client
+	 * connection had been closed. If it's a bug, then the correct
+	 * order of responses to requests may be broken.
+	 */
+	if (list_empty(seq_queue)) {
+		spin_unlock_bh(&cli_conn->msg_qlock);
+		req->flags |= TFW_HTTP_CONN_CLOSE;
+		tfw_http_conn_msg_free((TfwHttpMsg *)resp);
+		tfw_http_conn_cli_dropfree((TfwHttpMsg *)req);
+		TFW_INC_STAT_BH(serv.msgs_otherr);
+		return;
+	}
+	/*
+	 * Starting with the first request on the list, pick consecutive
+	 * requests that have a paired response. Remove those requests
+	 * from the list, and put them into the list of outgoing responses.
+	 */
+	do {
+		req = list_first_entry(seq_queue, TfwHttpReq, msg.seq_list);
+		if (req->resp == NULL)
+			break;
+		list_move_tail(&req->msg.seq_list, &out_queue);
+	} while(!list_empty(seq_queue));
+	spin_unlock_bh(&cli_conn->msg_qlock);
+
+	/* Forward responses to the client. */
+	list_for_each_entry_safe(req, tmp, &out_queue, msg.seq_list) {
+		list_del_init(&req->msg.seq_list);
+		resp = (TfwHttpResp *)req->resp;
+		/*
+		 * Verify that the client connection is not closing.
+		 * If it is, then discard @req and @resp, and return.
+		 * Remaining requests from the client will be taken
+		 * care of when the connection is released.
+		 */
+		if (!tfw_connection_nfo(cli_conn))
```

Comment:
"nfo" means "Not FailOvering". Since client connections can't be failovered, but now we call `tfw_connection_put_to_death()` for them, it's good to rename the function.